`used-before-assignment` false positive for `type` alias

### Bug description

```python
type IntOrX = int | X

class X:
    pass
```


### Configuration

_No response_

### Command used

```shell
pylint --py-version=3.12 asdf.py
```


### Pylint output

```python
asdf.py:1:20: E0601: Using variable 'X' before assignment (used-before-assignment)
```


### Expected behavior

No warning

> The value of the type alias is evaluated in the annotation scope. It is not evaluated when the type alias is created, but only when the value is accessed through the type alias’s __value__ attribute (see [Lazy evaluation](https://docs.python.org/3/reference/executionmodel.html#lazy-evaluation)). This allows the type alias to refer to names that are not yet defined.

https://docs.python.org/3/reference/simple_stmts.html#the-type-statement

### Pylint version

```shell
pylint 3.3.0-dev0
astroid 3.3.0-dev0
Python 3.12.3
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9815.false_positive
+++ b/doc/whatsnew/fragments/9815.false_positive
@@ -1,0 +1,3 @@
+Fix used-before-assignment for PEP 695 type aliases and parameters.
+
+Closes #9815

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -1646,6 +1646,13 @@ def is_node_in_type_annotation_context(node: nodes.NodeNG) -> bool:
             return False
 
 
+def is_node_in_pep695_type_context(node: nodes.NodeNG) -> nodes.NodeNG | None:
+    """Check if node is used in a TypeAlias or as part of a type param."""
+    return get_node_first_ancestor_of_type(
+        node, (nodes.TypeAlias, nodes.TypeVar, nodes.ParamSpec, nodes.TypeVarTuple)
+    )
+
+
 def is_subclass_of(child: nodes.ClassDef, parent: nodes.ClassDef) -> bool:
     """Check if first node is a subclass of second node.
 

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2021,12 +2021,12 @@ class VariablesChecker(BaseChecker):
 
         Returns True if an error is reported; otherwise, returns False.
         """
-        if self._is_builtin(node.name):
-            return False
         if (
             self._postponed_evaluation_enabled
             and utils.is_node_in_type_annotation_context(node)
         ) or utils.is_node_in_pep695_type_context(node):
+            return False
+        if self._is_builtin(node.name):
             return False
         if self._is_variable_annotation_in_function(node):
             return False

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1927,22 +1927,28 @@ class VariablesChecker(BaseChecker):
 
                 # Skip postponed evaluation of annotations
                 # and unevaluated annotations inside a function body
-                if not (
-                    self._postponed_evaluation_enabled
-                    and (
-                        isinstance(stmt, nodes.AnnAssign)
-                        or (
-                            isinstance(stmt, nodes.FunctionDef)
-                            and node
-                            not in {
-                                *(stmt.args.defaults or ()),
-                                *(stmt.args.kw_defaults or ()),
-                            }
+                if (
+                    not (
+                        self._postponed_evaluation_enabled
+                        and (
+                            isinstance(stmt, nodes.AnnAssign)
+                            or (
+                                isinstance(stmt, nodes.FunctionDef)
+                                and node
+                                not in {
+                                    *(stmt.args.defaults or ()),
+                                    *(stmt.args.kw_defaults or ()),
+                                }
+                            )
                         )
                     )
-                ) and not (
-                    isinstance(stmt, nodes.AnnAssign)
-                    and utils.get_node_first_ancestor_of_type(stmt, nodes.FunctionDef)
+                    and not (
+                        isinstance(stmt, nodes.AnnAssign)
+                        and utils.get_node_first_ancestor_of_type(
+                            stmt, nodes.FunctionDef
+                        )
+                    )
+                    and not isinstance(stmt, nodes.TypeAlias)
                 ):
                     self.add_message(
                         "used-before-assignment",
@@ -2015,12 +2021,12 @@ class VariablesChecker(BaseChecker):
 
         Returns True if an error is reported; otherwise, returns False.
         """
+        if self._is_builtin(node.name):
+            return False
         if (
             self._postponed_evaluation_enabled
             and utils.is_node_in_type_annotation_context(node)
-        ):
-            return False
-        if self._is_builtin(node.name):
+        ) or utils.is_node_in_pep695_type_context(node):
             return False
         if self._is_variable_annotation_in_function(node):
             return False

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1927,28 +1927,21 @@ class VariablesChecker(BaseChecker):
 
                 # Skip postponed evaluation of annotations
                 # and unevaluated annotations inside a function body
-                if (
-                    not (
-                        self._postponed_evaluation_enabled
-                        and (
-                            isinstance(stmt, nodes.AnnAssign)
-                            or (
-                                isinstance(stmt, nodes.FunctionDef)
-                                and node
-                                not in {
-                                    *(stmt.args.defaults or ()),
-                                    *(stmt.args.kw_defaults or ()),
-                                }
-                            )
-                        )
-                    )
-                    and not (
+                # as well as TypeAlias nodes.
+                if not (
+                    self._postponed_evaluation_enabled  # noqa: RUF021
+                    and (
                         isinstance(stmt, nodes.AnnAssign)
-                        and utils.get_node_first_ancestor_of_type(
-                            stmt, nodes.FunctionDef
-                        )
+                        or isinstance(stmt, nodes.FunctionDef)  # noqa: RUF021
+                        and node
+                        not in {
+                            *(stmt.args.defaults or ()),
+                            *(stmt.args.kw_defaults or ()),
+                        }
                     )
-                    and not isinstance(stmt, nodes.TypeAlias)
+                    or isinstance(stmt, nodes.AnnAssign)  # noqa: RUF021
+                    and utils.get_node_first_ancestor_of_type(stmt, nodes.FunctionDef)
+                    or isinstance(stmt, nodes.TypeAlias)
                 ):
                     self.add_message(
                         "used-before-assignment",

--- a/pylint/testutils/functional/find_functional_tests.py
+++ b/pylint/testutils/functional/find_functional_tests.py
@@ -21,6 +21,7 @@ IGNORED_PARENT_DIRS = {
     "ext",
     "regression",
     "regression_02",
+    "used_02",
 }
 """Direct parent directories that should be ignored."""
 

--- a/tests/functional/u/used/used_before_assignment_py312.py
+++ b/tests/functional/u/used/used_before_assignment_py312.py
@@ -1,14 +1,23 @@
-"""used-before-assignment re: python 3.12 generic typing syntax (PEP 695)"""
+"""Tests for used-before-assignment with Python 3.12 generic typing syntax (PEP 695)"""
+# pylint: disable = invalid-name,missing-docstring,too-few-public-methods,unused-argument
 
-from typing import Callable
+from typing import TYPE_CHECKING, Callable
+
 type Point[T] = tuple[T, ...]
 type Alias[*Ts] = tuple[*Ts]
-type Alias[**P] = Callable[P]
+type Alias2[**P] = Callable[P, None]
 
-# pylint: disable = invalid-name, missing-class-docstring, too-few-public-methods
-
-# https://github.com/pylint-dev/pylint/issues/9815
-type IntOrX = int | X  # [used-before-assignment] FALSE POSITIVE
+type AliasType = int | X | Y
 
 class X:
     pass
+
+if TYPE_CHECKING:
+    class Y: ...
+
+class Good[T: Y]: ...
+type OtherAlias[T: Y] = T | None
+
+# https://github.com/pylint-dev/pylint/issues/9884
+def func[T: Y](x: T) -> None:  # [redefined-outer-name]  FALSE POSITIVE
+    ...

--- a/tests/functional/u/used/used_before_assignment_py312.txt
+++ b/tests/functional/u/used/used_before_assignment_py312.txt
@@ -1,1 +1,1 @@
-used-before-assignment:11:20:11:21::Using variable 'X' before assignment:HIGH
+redefined-outer-name:22:9:22:13:func:Redefining name 'T' from outer scope (line 6):UNDEFINED

--- a/tests/functional/u/used_02/used_before_assignment_py313.py
+++ b/tests/functional/u/used_02/used_before_assignment_py313.py
@@ -1,0 +1,16 @@
+"""Tests for used-before-assignment with Python 3.13 type var defaults (PEP 696)"""
+# pylint: disable=missing-docstring,unused-argument,too-few-public-methods
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    class Y: ...
+
+class Good1[T = Y]: ...
+class Good2[*Ts = tuple[int, Y]]: ...
+class Good3[**P = [int, Y]]: ...
+type Alias[T = Y] = T | None
+
+# https://github.com/pylint-dev/pylint/issues/9884
+def func[T = Y](x: T) -> None:  # [redefined-outer-name]  FALSE POSITIVE
+    ...

--- a/tests/functional/u/used_02/used_before_assignment_py313.rc
+++ b/tests/functional/u/used_02/used_before_assignment_py313.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.13

--- a/tests/functional/u/used_02/used_before_assignment_py313.txt
+++ b/tests/functional/u/used_02/used_before_assignment_py313.txt
@@ -1,0 +1,1 @@
+redefined-outer-name:15:9:15:14:func:Redefining name 'T' from outer scope (line 12):UNDEFINED

